alica_engine/include/engine/blackboard/Blackboard.h

```diff
@@ -1,47 +1,325 @@
+    static constexpr std::size_t BB_VALUE_TYPE_ANY_INDEX = 5;
+
     template <class... Args>
-    void registerValue(const std::string& key, Args&&... args)
+    static std::string stringify(Args&&... args)
     {
-        vals.emplace(std::piecewise_construct, std::forward_as_tuple(key), std::forward_as_tuple(std::forward<decltype(args)>(args)...));
+        std::ostringstream oss;
+        (oss << ... << std::forward<Args>(args));
+        return oss.str();
+    }
+
+public:
+    BlackboardImpl() = default;
+    BlackboardImpl(const BlackboardImpl&) = delete;
+    BlackboardImpl(const BlackboardBlueprint* blueprint)
+    {
+        for (auto it = blueprint->begin(); it != blueprint->end(); ++it) {
+            const auto& [key, keyInfo] = *it;
+            auto typeIndex = getTypeIndex(keyInfo.type);
+            if (!typeIndex.has_value() && keyInfo.type != "std::any") {
+                throw BlackboardException(stringify("key: ", key, " has an unsupported type"));
+            }
+            _yamlType.emplace(std::piecewise_construct, std::forward_as_tuple(key), std::forward_as_tuple(keyInfo.type));
+            if (keyInfo.defaultValue.has_value()) {
+                if (keyInfo.type == "std::any") {
+                    throw BlackboardException(stringify("key: ", key, " cannot have a default value because it is of type std::any"));
+                }
+                try {
+                    _vals.emplace(std::piecewise_construct, std::forward_as_tuple(key),
+                            std::forward_as_tuple(makeBBValueForIndex<true>::make(typeIndex.value(), keyInfo.defaultValue.value())));
+                } catch (const std::exception& ex) {
+                    throw BlackboardException(stringify("Could not set key: ", key, ", from default value: ", keyInfo.defaultValue.value(),
+                            ", type: ", keyInfo.type, ", details: ", ex.what()));
+                }
+            } else {
+                if (keyInfo.type == "std::any") {
+                    // explicitly set the index for std::any, since getTypeIndex() returns empty for std::any
+                    typeIndex = BB_VALUE_TYPE_ANY_INDEX;
+                }
+                try {
+                    // insert a default constructed value
+                    _vals.emplace(
+                            std::piecewise_construct, std::forward_as_tuple(key), std::forward_as_tuple(makeBBValueForIndex<false>::make(typeIndex.value())));
+                } catch (const std::exception& ex) {
+                    throw BlackboardException(stringify("Could not initialize key: ", key, " with value of type: ", keyInfo.type, ", details: ", ex.what()));
+                }
+            }
+        }
     }
 
     template <typename T>
     const T& get(const std::string& key) const
     {
-        return std::any_cast<const T&>(vals.at(key));
+        // T must be an exact match to the T used while setting this key (i.e. no type conversions taken into account)
```

Comment:
this comment is slightly inaccurate for keys that are not in the yaml file & whose type does not match exactly with any type in the variant. More precisely, T should be an exact match to the type stored in the variant.